File: backend/app/best_times.py
"""Parse a Lenex results .lxf and populate best times."""
from __future__ import annotations

import re
import zipfile
from datetime import date as _date
from io import BytesIO
from xml.etree import ElementTree as ET

import json as _json

from sqlalchemy.orm import Session
from .models import AppConfig, Athlete, BestTime, Club, Gender


def _lenex_time_to_ms(t: str) -> int | None:
    """Convert Lenex time 'HH:MM:SS.hh' or 'MM:SS.hh' to ms."""
    if not t or t == "NT":
        return None
    m = re.match(r"(\d+):(\d+):(\d+)\.(\d+)", t)
    if m:
        return (int(m.group(1)) * 3600000 + int(m.group(2)) * 60000
                + int(m.group(3)) * 1000 + int(m.group(4)) * 10)
    m = re.match(r"(\d+):(\d+)\.(\d+)", t)
    if m:
        return (int(m.group(1)) * 60000 + int(m.group(2)) * 1000
                + int(m.group(3)) * 10)
    m = re.match(r"(\d+)\.(\d+)", t)
    if m:
        return int(m.group(1)) * 1000 + int(m.group(2)) * 10
    return None


def _find_or_create_athlete(db: Session, first: str, last: str, license: str, club=None) -> Athlete | None:
    """Match athlete by license first, then name. Create if not found and club provided."""
    if license:
        ath = db.query(Athlete).filter(Athlete.license == license).first()
        if ath:
            return ath
    ath = db.query(Athlete).filter(
        Athlete.first_name == first, Athlete.last_name == last
    ).first()
    if ath:
        return ath
    if not club:
        return None
    ath = Athlete(first_name=first, last_name=last, gender=Gender.M, club_id=club.id, license=license)
    db.add(ath)
    db.flush()
    return ath


def _upsert_best_time(db: Session, athlete_id: int, style_uid: int,
                      time_ms: int, course: str, source: str,
                      recorded_on: _date | None = None) -> bool:
    """Upsert a BT row, only overwriting if the new time is faster.
    Returns True when a row was inserted or improved.
    recorded_on is synced to the sibling course row so both LCM/SCM share one date."""
    existing = db.query(BestTime).filter(
        BestTime.athlete_id == athlete_id,
        BestTime.style_uid == style_uid,
        BestTime.course == course,
    ).first()
    if existing:
        if time_ms < existing.time_ms:
            existing.time_ms = time_ms
            existing.source = source
            if recorded_on is not None:
                existing.recorded_on = recorded_on
            improved = True
        else:
            improved = False
    else:
        db.add(BestTime(
            athlete_id=athlete_id,
            style_uid=style_uid,
            time_ms=time_ms,
            course=course,
            source=source,
            recorded_on=recorded_on,
        ))
        improved = True

    # Sync recorded_on to the sibling course row so LCM and SCM share one date
    if recorded_on is not None:
        sibling_course = "SCM" if course == "LCM" else "LCM"
        sibling = db.query(BestTime).filter(
            BestTime.athlete_id == athlete_id,
            BestTime.style_uid == style_uid,
            BestTime.course == sibling_course,
        ).first()
        if sibling:
            sibling.recorded_on = recorded_on

    return improved


def load_best_times(db: Session, file_bytes: bytes, source: str = "") -> dict:
    """Parse results .lxf and upsert best times. Returns counts."""
    with zipfile.ZipFile(BytesIO(file_bytes)) as z:
        lef_name = [n for n in z.namelist() if n.endswith(".lef")][0]
        xml_bytes = z.read(lef_name)

    root = ET.fromstring(xml_bytes)

    # Get course and date from MEET element
    meet_el = root.find(".//MEET")
    course = meet_el.get("course", "LCM") if meet_el is not None else "LCM"
    if course not in ("LCM", "SCM"):
        course = "LCM"  # treat SCY etc. as LCM
    recorded_on: _date | None = None
    if meet_el is not None:
        for date_attr in ("startdate", "date"):
            raw = meet_el.get(date_attr, "")
            if raw:
                try:
                    recorded_on = _date.fromisoformat(raw[:10])
                except ValueError:
                    pass
                if recorded_on:
                    break

    # Build eventid -> style_uid map and uid -> name from the Lenex events
    event_style: dict[str, int] = {}
    style_names: dict[int, str] = {}
    for event_el in root.iter("EVENT"):
        eid = event_el.get("eventid", "")
        for ss in event_el.iter("SWIMSTYLE"):
            uid_raw = ss.get("swimstyleid") or ss.get("stroke", "")
            try:
                uid_int = int(uid_raw)
            except (ValueError, TypeError):
                continue
            event_style[eid] = uid_int
            name = ss.get("name", "")
            if name and uid_int not in style_names:
                style_names[uid_int] = name

    updated = 0
    skipped = 0
    athletes_created = 0
    # Map Lenex athleteid -> DB Athlete, used later to attribute relay times.
    athlete_by_lenex_id: dict[str, Athlete] = {}

    for club_el in root.iter("CLUB"):
        club_name = club_el.get("name", "")
        club = db.query(Club).filter(Club.name == club_name).first()
        for ath_el in club_el.iter("ATHLETE"):
            first = ath_el.get("firstname", "")
            last = ath_el.get("lastname", "")
            license = ath_el.get("license", "")
            gender_str = ath_el.get("gender", "M")
            bd_str = ath_el.get("birthdate", "")
            lenex_aid = ath_el.get("athleteid", "")

            athlete = _find_or_create_athlete(db, first, last, license, club)
            if not athlete:
                skipped += 1
                continue
            if lenex_aid:
                athlete_by_lenex_id[lenex_aid] = athlete
            # Update gender/birthdate if newly created
            if athlete.id is None or (not athlete.birthdate and bd_str):
                athlete.gender = Gender.F if gender_str == "F" else Gender.M
                if bd_str:
                    try:
                        athlete.birthdate = _date.fromisoformat(bd_str)
                    except ValueError:
                        pass
                athletes_created += 1

            # Collect best candidate times per (event, course) pair.
            # entrycourse on individual ENTRY elements overrides the meet-level course
            # so that a multi-course export/backup round-trips correctly.
            event_times: dict[tuple[str, str], list[int]] = {}
            for entry_el in ath_el.iter("ENTRY"):
                eid = entry_el.get("eventid", "")
                t = _lenex_time_to_ms(entry_el.get("entrytime", ""))
                if t and eid:
                    ec = entry_el.get("entrycourse", "") or course
                    if ec not in ("LCM", "SCM"):
                        ec = course
                    event_times.setdefault((eid, ec), []).append(t)
            for result_el in ath_el.iter("RESULT"):
                eid = result_el.get("eventid", "")
                t = _lenex_time_to_ms(result_el.get("swimtime", ""))
                if t and eid:
                    event_times.setdefault((eid, course), []).append(t)

            for (eid, ev_course), times in event_times.items():
                style_uid = event_style.get(eid)
                if not style_uid:
                    continue
                if _upsert_best_time(db, athlete.id, style_uid,
                                     min(times), ev_course, source, recorded_on):
                    updated += 1

    # Relay BT: each member of a team gets the team time recorded against the
    # relay's style_uid, mirroring how individual times update each athlete's BT.
    for relay_el in root.iter("RELAY"):
        # The roster (RELAYPOSITIONS) is shared across all entries/results
        # of a given <RELAY>. SPLASH only writes it on the first one and
        # omits it from siblings, so fall back to the first list we find.
        roster: list[Athlete] = []
        for pos_el in relay_el.iter("RELAYPOSITION"):
            ath = athlete_by_lenex_id.get(pos_el.get("athleteid", ""))
            if ath and ath not in roster:
                roster.append(ath)
        if not roster:
            continue

        relay_event_times: dict[str, list[int]] = {}
        for entry_el in relay_el.iter("ENTRY"):
            eid = entry_el.get("eventid", "")
            t = _lenex_time_to_ms(entry_el.get("entrytime", ""))
            if t and eid:
                relay_event_times.setdefault(eid, []).append(t)
        for result_el in relay_el.iter("RESULT"):
            eid = result_el.get("eventid", "")
            t = _lenex_time_to_ms(result_el.get("swimtime", ""))
            if t and eid:
                relay_event_times.setdefault(eid, []).append(t)

        for eid, times in relay_event_times.items():
            style_uid = event_style.get(eid)
            if not style_uid:
                continue
            best = min(times)
            for ath in roster:
                if _upsert_best_time(db, ath.id, style_uid, best, course, source, recorded_on):
                    updated += 1

    # Persist style uid→name so the Data Management page can show names even without a meet file.
    if style_names:
        cfg = db.query(AppConfig).get("style_names_json")
        existing: dict[int, str] = _json.loads(cfg.value) if cfg else {}
        # Convert keys to int for merge; new names win if uid not already known
        merged = {int(k): v for k, v in existing.items()}
        for uid, name in style_names.items():
            if uid not in merged:
                merged[uid] = name
        payload = _json.dumps({str(k): v for k, v in merged.items()})
        if cfg:
            cfg.value = payload
        else:
            db.add(AppConfig(key="style_names_json", value=payload))

    db.commit()
    return {"times_updated": updated, "athletes_skipped": skipped, "athletes_created": athletes_created}

File: backend/src/best_times.rs
use chrono::NaiveDate;
use quick_xml::events::Event as XmlEvent;
use quick_xml::Reader;
use serde::Serialize;
use sqlx::PgPool;
use std::collections::HashMap;

use crate::meet_parser::extract_lef_from_zip;

#[derive(Debug, Serialize)]
pub struct BestTimesResult {
    pub times_updated: i32,
    pub athletes_skipped: i32,
    pub athletes_created: i32,
}

/// Convert Lenex time string to milliseconds
fn lenex_time_to_ms(t: &str) -> Option<i32> {
    if t.is_empty() || t == "NT" {
        return None;
    }
    // HH:MM:SS.hh
    if let Some(caps) = regex_match_4(t) {
        return Some(caps.0 * 3600000 + caps.1 * 60000 + caps.2 * 1000 + caps.3 * 10);
    }
    // MM:SS.hh
    if let Some(caps) = regex_match_3(t) {
        return Some(caps.0 * 60000 + caps.1 * 1000 + caps.2 * 10);
    }
    // SS.hh
    if let Some(caps) = regex_match_2(t) {
        return Some(caps.0 * 1000 + caps.1 * 10);
    }
    None
}

fn regex_match_4(t: &str) -> Option<(i32, i32, i32, i32)> {
    let parts: Vec<&str> = t.splitn(2, '.').collect();
    if parts.len() != 2 {
        return None;
    }
    let cs: i32 = parts[1].parse().ok()?;
    let time_parts: Vec<&str> = parts[0].split(':').collect();
    if time_parts.len() != 3 {
        return None;
    }
    Some((
        time_parts[0].parse().ok()?,
        time_parts[1].parse().ok()?,
        time_parts[2].parse().ok()?,
        cs,
    ))
}

fn regex_match_3(t: &str) -> Option<(i32, i32, i32)> {
    let parts: Vec<&str> = t.splitn(2, '.').collect();
    if parts.len() != 2 {
        return None;
    }
    let cs: i32 = parts[1].parse().ok()?;
    let time_parts: Vec<&str> = parts[0].split(':').collect();
    if time_parts.len() != 2 {
        return None;
    }
    Some((
        time_parts[0].parse().ok()?,
        time_parts[1].parse().ok()?,
        cs,
    ))
}

fn regex_match_2(t: &str) -> Option<(i32, i32)> {
    let parts: Vec<&str> = t.splitn(2, '.').collect();
    if parts.len() != 2 {
        return None;
    }
    Some((parts[0].parse().ok()?, parts[1].parse().ok()?))
}

async fn upsert_best_time(
    pool: &PgPool,
    athlete_id: i32,
    style_uid: i32,
    time_ms: i32,
    course: &str,
    source: &str,
    recorded_on: Option<NaiveDate>,
) -> Result<bool, sqlx::Error> {
    let existing: Option<(i32, i32, Option<NaiveDate>)> = sqlx::query_as(
        "SELECT id, time_ms, recorded_on FROM best_times WHERE athlete_id = $1 AND style_uid = $2 AND course = $3",
    )
    .bind(athlete_id)
    .bind(style_uid)
    .bind(course)
    .fetch_optional(pool)
    .await?;

    let improved = if let Some((id, existing_ms, existing_date)) = existing {
        if time_ms < existing_ms {
            sqlx::query("UPDATE best_times SET time_ms = $1, source = $2, recorded_on = COALESCE($3, recorded_on) WHERE id = $4")
                .bind(time_ms)
                .bind(source)
                .bind(recorded_on)
                .bind(id)
                .execute(pool)
                .await?;
            true
        } else {
            if recorded_on.is_some() && existing_date.is_none() {
                sqlx::query("UPDATE best_times SET recorded_on = $1 WHERE id = $2")
                    .bind(recorded_on)
                    .bind(id)
                    .execute(pool)
                    .await?;
            }
            false
        }
    } else {
        sqlx::query(
            "INSERT INTO best_times (athlete_id, style_uid, time_ms, course, source, recorded_on) VALUES ($1, $2, $3, $4, $5, $6)",
        )
        .bind(athlete_id)
        .bind(style_uid)
        .bind(time_ms)
        .bind(course)
        .bind(source)
        .bind(recorded_on)
        .execute(pool)
        .await?;
        true
    };

    // Sync recorded_on to sibling course
    if let Some(date) = recorded_on {
        let sibling_course = if course == "LCM" { "SCM" } else { "LCM" };
        sqlx::query("UPDATE best_times SET recorded_on = $1 WHERE athlete_id = $2 AND style_uid = $3 AND course = $4")
            .bind(date)
            .bind(athlete_id)
            .bind(style_uid)
            .bind(sibling_course)
            .execute(pool)
            .await?;
    }

    Ok(improved)
}

pub async fn load_best_times(pool: &PgPool, data: &[u8], source: &str) -> Result<BestTimesResult, String> {
    let xml = extract_lef_from_zip(data)?;
    let mut reader = Reader::from_reader(xml.as_slice());
    reader.config_mut().trim_text(true);
    let mut buf = Vec::new();

    let mut course = "LCM".to_string();
    let mut recorded_on: Option<NaiveDate> = None;
    let mut event_style: HashMap<String, i32> = HashMap::new();
    let mut updated = 0i32;
    let mut skipped = 0i32;

    // First pass: get meet-level course and date, and build event->style map
    // We'll do a single-pass approach with state tracking
    let mut in_club = false;
    let mut in_athlete = false;
    let mut in_event_def = false;
    let mut current_club_code = String::new();
    let mut current_club_name = String::new();
    let mut current_ath_first = String::new();
    let mut current_ath_last = String::new();
    let mut current_ath_license = String::new();

    // We need two passes: first to get event_style map, then to process athletes
    // Let's collect everything in one pass using state

    // Actually, let's do a simpler two-pass approach
    drop(reader);

    // Pass 1: extract metadata
    let mut reader = Reader::from_reader(xml.as_slice());
    reader.config_mut().trim_text(true);
    loop {
        match reader.read_event_into(&mut buf) {
            Ok(XmlEvent::Eof) => break,
            Ok(XmlEvent::Start(ref e)) | Ok(XmlEvent::Empty(ref e)) => {
                let tag = String::from_utf8_lossy(e.name().as_ref()).to_string();
                match tag.as_str() {
                    "MEET" => {
                        for attr in e.attributes().flatten() {
                            let key = String::from_utf8_lossy(attr.key.as_ref()).to_string();
                            let val = String::from_utf8_lossy(&attr.value).to_string();
                            match key.as_str() {
                                "course" => {
                                    if val == "LCM" || val == "SCM" {
                                        course = val;
                                    }
                                }
                                "startdate" | "date" => {
                                    if recorded_on.is_none() {
                                        recorded_on = NaiveDate::parse_from_str(
                                            &val[..10.min(val.len())],
                                            "%Y-%m-%d",
                                        )
                                        .ok();
                                    }
                                }
                                _ => {}
                            }
                        }
                    }
                    "SESSION" => {
                        if recorded_on.is_none() {
                            for attr in e.attributes().flatten() {
                                let key = String::from_utf8_lossy(attr.key.as_ref()).to_string();
                                let val = String::from_utf8_lossy(&attr.value).to_string();
                                if key == "date" {
                                    recorded_on = NaiveDate::parse_from_str(
                                        &val[..10.min(val.len())],
                                        "%Y-%m-%d",
                                    )
                                    .ok();
                                }
                            }
                        }
                    }
                    "EVENT" => {
                        let mut eid = String::new();
                        for attr in e.attributes().flatten() {
                            let key = String::from_utf8_lossy(attr.key.as_ref()).to_string();
                            let val = String::from_utf8_lossy(&attr.value).to_string();
                            if key == "eventid" {
                                eid = val;
                            }
                        }
                        if !eid.is_empty() {
                            // Look for SWIMSTYLE in next events
                            // Store eid for when we find SWIMSTYLE
                            current_club_code = eid; // reuse variable temporarily
                            in_event_def = true;
                        }
                    }
                    "SWIMSTYLE" if in_event_def => {
                        for attr in e.attributes().flatten() {
                            let key = String::from_utf8_lossy(attr.key.as_ref()).to_string();
                            let val = String::from_utf8_lossy(&attr.value).to_string();
                            if key == "swimstyleid" {
                                if let Ok(uid) = val.parse::<i32>() {
                                    event_style.insert(current_club_code.clone(), uid);
                                }
                            }
                        }
                    }
                    _ => {}
                }
            }
            Ok(XmlEvent::End(ref e)) => {
                let tag = String::from_utf8_lossy(e.name().as_ref()).to_string();
                if tag == "EVENT" {
                    in_event_def = false;
                }
            }
            Err(e) => return Err(format!("XML error: {e}")),
            _ => {}
        }
        buf.clear();
    }

    if recorded_on.is_none() {
        recorded_on = Some(chrono::Local::now().date_naive());
    }

    // Pass 2: process athletes and their times
    buf.clear();
    let mut reader = Reader::from_reader(xml.as_slice());
    reader.config_mut().trim_text(true);
    in_club = false;
    in_athlete = false;
    current_club_code = String::new();
    current_club_name = String::new();

    #[derive(Default)]
    struct AthEntries {
        first: String,
        last: String,
        license: String,
        // (eventid, course, time_ms, date)
        times: Vec<(String, String, i32, Option<NaiveDate>)>,
    }
    let mut current_ath = AthEntries::default();

    loop {
        match reader.read_event_into(&mut buf) {
            Ok(XmlEvent::Eof) => break,
            Ok(XmlEvent::Start(ref e)) | Ok(XmlEvent::Empty(ref e)) => {
                let tag = String::from_utf8_lossy(e.name().as_ref()).to_string();
                match tag.as_str() {
                    "CLUB" => {
                        in_club = true;
                        current_club_code = String::new();
                        current_club_name = String::new();
                        for attr in e.attributes().flatten() {
                            let key = String::from_utf8_lossy(attr.key.as_ref()).to_string();
                            let val = String::from_utf8_lossy(&attr.value).to_string();
                            match key.as_str() {
                                "code" => current_club_code = val,
                                "name" => current_club_name = val,
                                _ => {}
                            }
                        }
                    }
                    "ATHLETE" if in_club => {
                        in_athlete = true;
                        current_ath = AthEntries::default();
                        for attr in e.attributes().flatten() {
                            let key = String::from_utf8_lossy(attr.key.as_ref()).to_string();
                            let val = String::from_utf8_lossy(&attr.value).to_string();
                            match key.as_str() {
                                "firstname" => current_ath.first = val,
                                "lastname" => current_ath.last = val,
                                "license" => current_ath.license = val,
                                _ => {}
                            }
                        }
                    }
                    "ENTRY" if in_athlete => {
                        let mut eid = String::new();
                        let mut etime = String::new();
                        let mut ecourse = course.clone();
                        for attr in e.attributes().flatten() {
                            let key = String::from_utf8_lossy(attr.key.as_ref()).to_string();
                            let val = String::from_utf8_lossy(&attr.value).to_string();
                            match key.as_str() {
                                "eventid" => eid = val,
                                "entrytime" => etime = val,
                                "entrycourse" => {
                                    if val == "LCM" || val == "SCM" {
                                        ecourse = val;
                                    }
                                }
                                _ => {}
                            }
                        }
                        if let Some(ms) = lenex_time_to_ms(&etime) {
                            if !eid.is_empty() {
                                current_ath.times.push((eid, ecourse, ms, None));
                            }
                        }
                    }
                    "RESULT" if in_athlete => {
                        let mut eid = String::new();
                        let mut stime = String::new();
                        for attr in e.attributes().flatten() {
                            let key = String::from_utf8_lossy(attr.key.as_ref()).to_string();
                            let val = String::from_utf8_lossy(&attr.value).to_string();
                            match key.as_str() {
                                "eventid" => eid = val,
                                "swimtime" => stime = val,
                                _ => {}
                            }
                        }
                        if let Some(ms) = lenex_time_to_ms(&stime) {
                            if !eid.is_empty() {
                                current_ath.times.push((eid, course.clone(), ms, recorded_on));
                            }
                        }
                    }
                    _ => {}
                }
            }
            Ok(XmlEvent::End(ref e)) => {
                let tag = String::from_utf8_lossy(e.name().as_ref()).to_string();
                match tag.as_str() {
                    "ATHLETE" if in_athlete => {
                        // Process this athlete's times
                        let ath_id = find_athlete(pool, &current_ath.first, &current_ath.last, &current_ath.license).await;
                        if let Some(aid) = ath_id {
                            // Group by (eventid, course) and take best
                            let mut best_per_event: HashMap<(String, String), (i32, Option<NaiveDate>)> = HashMap::new();
                            for (eid, ec, ms, date) in &current_ath.times {
                                let key = (eid.clone(), ec.clone());
                                let entry = best_per_event.entry(key).or_insert((*ms, *date));
                                if *ms < entry.0 {
                                    *entry = (*ms, *date);
                                }
                            }
                            for ((eid, ec), (ms, date)) in &best_per_event {
                                if let Some(&uid) = event_style.get(eid) {
                                    let d = date.or(recorded_on);
                                    match upsert_best_time(pool, aid, uid, *ms, ec, source, d).await {
                                        Ok(true) => updated += 1,
                                        Ok(false) => {}
                                        Err(_) => {}
                                    }
                                }
                            }
                        } else {
                            skipped += 1;
                        }
                        in_athlete = false;
                    }
                    "CLUB" => in_club = false,
                    _ => {}
                }
            }
            Err(e) => return Err(format!("XML error: {e}")),
            _ => {}
        }
        buf.clear();
    }

    Ok(BestTimesResult {
        times_updated: updated,
        athletes_skipped: skipped,
        athletes_created: 0,
    })
}

async fn find_athlete(pool: &PgPool, first: &str, last: &str, license: &str) -> Option<i32> {
    if !license.is_empty() {
        let row: Option<(i32,)> = sqlx::query_as("SELECT id FROM athletes WHERE license = $1")
            .bind(license)
            .fetch_optional(pool)
            .await
            .ok()?;
        if let Some((id,)) = row {
            return Some(id);
        }
    }
    let row: Option<(i32,)> =
        sqlx::query_as("SELECT id FROM athletes WHERE first_name = $1 AND last_name = $2")
            .bind(first)
            .bind(last)
            .fetch_optional(pool)
            .await
            .ok()?;
    row.map(|(id,)| id)
}
